File: KisaragiTech.Dape/User/Model/HashedPassword.cs
using System;
using System.Globalization;
using System.Linq;
using System.Security.Cryptography;
using KisaragiTech.Dape.Base.Strings;
using Konscious.Security.Cryptography;

namespace KisaragiTech.Dape.User.Model;

public sealed record HashedPassword(ReadOnlyMemory<byte> Hash, ReadOnlyMemory<byte> Salt, int Iterations, int MemorySize, int DegreeOfParallelism)
{
    public bool Equals(HashedPassword? other) => other != null && CryptographicOperations.FixedTimeEquals(this.Hash.Span, other.Hash.Span);

    public override int GetHashCode() => System.HashCode.Combine(this.Hash);

    public string ToSerializationFormat() => Convert.ToBase64String(this.Hash.Span) + ":" + Convert.ToBase64String(this.Salt.Span) + ":" + this.Iterations + ":" + this.MemorySize + ":" + this.DegreeOfParallelism;

    public static HashedPassword Deserialize(string serialized)
    {
        var parts = StringSplitter.ToMemories(serialized, ':').ToList();
        return parts switch
        {
            [var hash, var salt, var iterations, var memorySize, var degreeOfParallelism] => new HashedPassword(
                Convert.FromBase64String(hash.ToString()),
                Convert.FromBase64String(salt.ToString()),
                int.Parse(iterations.Span, NumberStyles.Integer, CultureInfo.InvariantCulture),
                int.Parse(memorySize.Span, NumberStyles.Integer, CultureInfo.InvariantCulture),
                int.Parse(degreeOfParallelism.Span, NumberStyles.Integer, CultureInfo.InvariantCulture)),
            _ => throw new InvalidOperationException("Hashed password must contain 5 parts exactly."),
        };
    }

    public bool Verify(string password)
    {
        var salt = this.Salt.ToArray();
        using var argon2 = new Argon2id(System.Text.Encoding.UTF8.GetBytes(password))
        {
            Salt = salt,
            Iterations = this.Iterations,
            MemorySize = this.MemorySize,
            DegreeOfParallelism = this.DegreeOfParallelism,
        };
        var computed = argon2.GetBytes(this.Hash.Length);
        return CryptographicOperations.FixedTimeEquals(computed, this.Hash.Span);
    }
}

File: src/main/scala/io/github/ksrgtech/dape/user/model/HashedPassword.scala
package io.github.ksrgtech.dape.user.model

import cats.effect.kernel.Sync
import com.password4j.Argon2Function
import io.github.ksrgtech.dape.base.strings.StringSplitter

import java.security.MessageDigest
import java.util.Base64
import scala.util.Try

final case class HashedPassword(
    hash: Array[Byte],
    salt: Array[Byte],
    iterations: Int,
    memorySize: Int,
    degreeOfParallelism: Int
)

object HashedPassword {
  private val HashLength: Int = 128
  private val Base64Pattern   = "^[A-Za-z0-9+/]*={0,2}$".r

  private def decodeBase64Strict(input: String): Either[IllegalArgumentException, Array[Byte]] =
    // Check that the string is valid base64 format
    // Base64 encoded strings should have length that is a multiple of 4
    if input.isEmpty then {
      Left(new IllegalArgumentException("Invalid base64: empty string"))
    } else if input.length % 4 != 0 then {
      Left(new IllegalArgumentException(s"Invalid base64: length ${input.length} is not a multiple of 4"))
    } else if !Base64Pattern.matches(input) then {
      Left(new IllegalArgumentException("Invalid base64: contains invalid characters"))
    } else {
      Try(Base64.getDecoder.decode(input)).toEither.left.map { e =>
        new IllegalArgumentException(s"Invalid base64: ${e.getMessage}")
      }
    }

  def apply(
      hash: Array[Byte],
      salt: Array[Byte],
      iterations: Int,
      memorySize: Int,
      degreeOfParallelism: Int
  ): HashedPassword =
    HashedPasswordData(hash, salt, iterations, memorySize, degreeOfParallelism)

  def deserialize(serialized: String): Either[IllegalArgumentException, HashedPassword] = {
    val parts = StringSplitter.toSubstrings(serialized, ':').toList
    parts match {
      case hashStr :: saltStr :: iterStr :: memStr :: parallelStr :: Nil =>
        for {
          hash <- decodeBase64Strict(hashStr).left.map { e =>
            new IllegalArgumentException(s"Invalid hash format: ${e.getMessage}")
          }
          salt <- decodeBase64Strict(saltStr).left.map { e =>
            new IllegalArgumentException(s"Invalid salt format: ${e.getMessage}")
          }
          iterations <- Try(iterStr.toInt).toEither.left.map { e =>
            new IllegalArgumentException(s"Invalid iterations format: ${e.getMessage}")
          }
          memorySize <- Try(memStr.toInt).toEither.left.map { e =>
            new IllegalArgumentException(s"Invalid memory size format: ${e.getMessage}")
          }
          parallelism <- Try(parallelStr.toInt).toEither.left.map { e =>
            new IllegalArgumentException(s"Invalid parallelism format: ${e.getMessage}")
          }
        } yield HashedPassword(hash, salt, iterations, memorySize, parallelism)
      case _ =>
        Left(new IllegalArgumentException("Hashed password must contain 5 parts exactly."))
    }
  }

  def unsafeDeserialize(serialized: String): HashedPassword =
    deserialize(serialized) match {
      case Right(hp) => hp
      case Left(e)   => throw e
    }

  extension (hp: HashedPassword) {
    def hash: Array[Byte]        = hp.hash
    def salt: Array[Byte]        = hp.salt
    def iterations: Int          = hp.iterations
    def memorySize: Int          = hp.memorySize
    def degreeOfParallelism: Int = hp.degreeOfParallelism

    def toSerializationFormat: String = {
      val encoder = Base64.getEncoder
      s"${encoder.encodeToString(hp.hash)}:${encoder.encodeToString(hp.salt)}:${hp.iterations}:${hp.memorySize}:${hp.degreeOfParallelism}"
    }

    def verify[F[_]: Sync](password: String): F[Boolean] =
      Sync[F].delay {
        val argon2 = Argon2Function.getInstance(
          hp.memorySize,
          hp.iterations,
          hp.degreeOfParallelism,
          HashLength,
          com.password4j.types.Argon2.ID
        )
        val computed = argon2.hash(password.getBytes("UTF-8"), hp.salt).getBytes
        MessageDigest.isEqual(computed, hp.hash)
      }

    def verifyUnsafe(password: String): Boolean = {
      val argon2 = Argon2Function.getInstance(
        hp.memorySize,
        hp.iterations,
        hp.degreeOfParallelism,
        HashLength,
        com.password4j.types.Argon2.ID
      )
      val computed = argon2.hash(password.getBytes("UTF-8"), hp.salt).getBytes
      MessageDigest.isEqual(computed, hp.hash)
    }

    def equalsPassword(other: HashedPassword): Boolean =
      MessageDigest.isEqual(hp.hash, other.hash)
  }
}
